compiler/resolution/resolveFunction.cpp

```diff
@@ -755,27 +751,36 @@ static void insertUnrefForArrayOrTupleReturn(FnSymbol* fn) {
     if (CallExpr* call = toCallExpr(se->parentExpr)) {
       if (call->isPrimitive(PRIM_MOVE) == true &&
           call->get(1)                 == se) {
-        Type* rhsType = call->get(2)->typeInfo();
-
-        bool arrayIsh = (rhsType->symbol->hasFlag(FLAG_ARRAY) ||
+        Expr* fromExpr = call->get(2);
+        Type* rhsType = fromExpr->typeInfo();
+
+        SymExpr* fromSe = toSymExpr(fromExpr);
+        bool domain = rhsType->symbol->hasFlag(FLAG_DOMAIN) &&
+                      fromSe != NULL &&
+                      isCallExprTemporary(fromSe->symbol()) &&
+                      isTemporaryFromNoCopyReturn(fromSe->symbol());
+        bool array = rhsType->symbol->hasFlag(FLAG_ARRAY);
+        bool arrayIsh = (array ||
                          rhsType->symbol->hasFlag(FLAG_ITERATOR_RECORD));
 
+        bool handleDomain = skipArray == false && domain;
         bool handleArray = skipArray == false && arrayIsh;
         bool handleTuple = skipTuple == false &&
                            isTupleContainingAnyReferences(rhsType);
 
-        // TODO: Should we check if the RHS is a symbol with
-        // 'no auto destroy' on it? If it is, then we'd be copying
-        // the RHS and it would never be destroyed...
-        if ((handleArray || handleTuple) && !isTypeExpr(call->get(2))) {
+        if ((handleArray || handleDomain || handleTuple) &&
+            !isTypeExpr(call->get(2))) {
+
+          FnSymbol* initCopyFn = getInitCopyDuringResolution(rhsType);
+          INT_ASSERT(initCopyFn);
 
           SET_LINENO(call);
           Expr*      rhs       = call->get(2)->remove();
-          VarSymbol* tmp       = newTemp("array_unref_ret_tmp", rhsType);
+          VarSymbol* tmp       = newTemp("copy_ret_tmp", rhsType);
           CallExpr*  initTmp   = new CallExpr(PRIM_MOVE,     tmp, rhs);
-          CallExpr*  unrefCall = new CallExpr("chpl__unref", tmp);
+          Symbol* definedConst = gFalse;
```

Comment:
I am not so sure. The case this is relevant for is something like this:

``` chapel
var A:[1..10] int;
writeln("A.domain definedConst=", A.domain._value.definedConst);

proc returnADomainCopy() {
  return A.domain; // copies to a non "view"
}

proc test1() {
  const D = returnADomainCopy();
  writeln("test1 D definedConst=", D._value.definedConst);
}
test1();

proc test2() {
  var D = returnADomainCopy();
  writeln("test2 D definedConst=", D._value.definedConst);
  // Could now do D = {1..2} e.g. ; so D's definedConst must not be true
}
test2();
```

The point is that nothing in `returnADomainCopy` impacts the const-ness of where that variable ends up.

While we might be able to get `test1` to print out definedConst=true; I'm pretty sure that we shouldn't have `test2` print out definedConst=true. I'm not sure how important this is to definedConst (since it's more or less an optimization) but either way I think it needs to be future work.

There might be similar issues with copy elision or out intents. I do not know if any of these cases will lead us to find bugs with the current definedConst handling.

Reply:
I poked at it and I think there is a bug with definedConst handling - I made issue #16354 for that.